compiler/rustc_mir_transform/src/coverage/mod.rs

```diff
@@ -346,20 +346,15 @@ fn get_body_span<'tcx>(
     let mut body_span = hir_body.value.span;
 
     if tcx.is_closure(def_id.to_def_id()) {
-        // If the MIR function is a closure, and if the closure body span
-        // starts from a macro, but it's content is not in that macro, try
-        // to find a non-macro callsite, and instrument the spans there
-        // instead.
-        loop {
-            let expn_data = body_span.ctxt().outer_expn_data();
-            if expn_data.is_root() {
-                break;
-            }
-            if let ExpnKind::Macro { .. } = expn_data.kind {
-                body_span = expn_data.call_site;
-            } else {
-                break;
-            }
+        // If the current function is a closure, and its "body" span was created
+        // by macro expansion or async desugaring, try to walk backwards to the
+        // pre-expansion call site or body.
+        while let expn_data = body_span.ctxt().outer_expn_data()
+            && !expn_data.is_root()
+            && let ExpnKind::Macro { .. } | ExpnKind::Desugaring(DesugaringKind::Async) =
+                expn_data.kind
+        {
+            body_span = expn_data.call_site;
```

Comment:
I'm quite pleased with how using a let-chain has greatly simplified the control flow of this loop.